cellbase-lib/src/main/java/org/opencb/cellbase/lib/impl/core/DataReleaseSingleton.java

```diff
@@ -0,0 +1,199 @@
+    // {key = dbname, value = lock}
+    private Map<String, ReentrantReadWriteLock> rwLockMap = new HashMap<>();
+
+    // {key = dbname, value = { key = release, value = { key = data, value = collection } } }
+    private Map<String, Map<Integer, Map<String, MongoDBCollection>>> cachedData = new HashMap<>();
+
+    private CellBaseManagerFactory managerFactory;
+
+    private static DataReleaseSingleton instance;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DataReleaseSingleton.class);
+
+    public static final String UNKOWN_DATABASE_MSG_PREFIX = "Unknown database ";
+    public static final String INVALID_RELEASE_MSG_PREFIX = "Invalid release ";
+    public static final String INVALID_DATA_MSG_PREFIX = "Invalid data ";
+
+    // Private constructor to prevent instantiation
+    private DataReleaseSingleton(CellBaseManagerFactory managerFactory) throws CellBaseException {
+        this.managerFactory = managerFactory;
+
+        // Support multi species and assemblies
+        CellBaseConfiguration configuration = managerFactory.getConfiguration();
+        for (SpeciesConfiguration vertebrate : configuration.getSpecies().getVertebrates()) {
+            for (SpeciesConfiguration.Assembly assembly : vertebrate.getAssemblies()) {
+                String databaseName = MongoDBManager.getDatabaseName(vertebrate.getId(), assembly.getName(), configuration.getVersion());
+                // This is necessary, before creating the database name the assembly is "cleaned", and we need to get the data release
+                // manager from the species and the assembly
+                speciesMap.put(databaseName, vertebrate.getId());
+                assemblyMap.put(databaseName, assembly.getName());
+                rwLockMap.put(databaseName, new ReentrantReadWriteLock());
+                cachedData.put(databaseName, new HashMap<>());
+
+                MongoClient mongoClient = managerFactory.getDataReleaseManager(vertebrate.getId(), assembly.getName()).getMongoDatastore()
+                        .getMongoClient();
+                MongoDatabase database = mongoClient.getDatabase(databaseName);
+                MongoCollection<Document> collection = database.getCollection(ReleaseMongoDBAdaptor.DATA_RELEASE_COLLECTION_NAME);
+                LOGGER.info("Setting listener for database {} and collection {}", database.getName(), collection.getNamespace()
+                        .getCollectionName());
+                // Set up the change stream for the collection
+                new Thread(() -> {
+                    collection.watch().fullDocument(FullDocument.UPDATE_LOOKUP).forEach(changeStreamDocument -> {
+                        try {
+                            handleDocumentChange(changeStreamDocument);
+                        } catch (CellBaseException e) {
+                            LOGGER.warn("Exception from handle document change function: {}", e.getStackTrace());
```

Comment:
Improper exception reporting. The whole exception should be the last element of the logger.
```
LOGGER.warn("Exception from handle document change function", e);
```
Also, might want to add more details on the collection watcher that failed.

Reply:
And what if the whole `collection.watch()...` fails? The whole code of the thread should be within a pseudo infinite loop, catching any `Exception`, and check for thread interruptions within the loop

```java
while (!Thread.currentThread().isInterrupted()) {
  try {
     MongoCursor iterator = collection.watch().fullDocument(FullDocument.UPDATE_LOOKUP).iterator();
     while (!Thread.currentThread().isInterrupted() && iterator.hasNext()) {
        // ...
     }
  } catch (Exception e) {
    LOGGER.warn("...", e);
  }
}
```

Reply:
> Improper exception reporting. The whole exception should be the last element of the logger.
> 
> ```
> LOGGER.warn("Exception from handle document change function", e);
> ```
> 
> Also, might want to add more details on the collection watcher that failed.

maybe it's better to display the strack trace, e.g.:
LOGGER.warn("Exception from handle document change function: {}", Arrays.toString(e.getStackTrace()));

Reply:
Having the exception as the last argument in a logger will print the whole stacktrace

```java
        try {
            throw new RuntimeException("myException");
        } catch (Exception e) {
            logger.warn("This is an exception", e);
        }
```

```
2024-08-08 12:03:27 [main] WARN  VariantQueryExecutorTest:59 - This is an exception
java.lang.RuntimeException: myException
	at org.opencb.opencga.storage.core.variant.query.executors.VariantQueryExecutorTest.setUp(VariantQueryExecutorTest.java:57) ~[test-classes/:?]
	at org.opencb.opencga.storage.hadoop.variant.query.executors.HadoopVariantQueryExecutorTest.setUp(HadoopVariantQueryExecutorTest.java:25) ~[test-classes/:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_161]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_161]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_161]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_161]
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56) ~[junit-4.13.2.jar:4.13.2]
	at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33) ~[junit-4.13.2.jar:4.13.2]
.....

```